packages/chart/src/Contour.ts

```diff
@@ -0,0 +1,92 @@
+import { SVGWidget } from "@hpcc-js/common";
+import { contourDensity as d3ContourDensity } from "d3-contour";
+import { geoPath } from "d3-geo";
+import { XYAxis } from "./XYAxis";
+
+export class Contour extends XYAxis {
+    constructor() {
+        super();
+        this
+            .xAxisGuideLines_default(false)
+            .yAxisGuideLines_default(false)
+            ;
+    }
+
+    xPos(host: XYAxis, d) {
+        return host.orientation() === "horizontal" ? host.dataPos(d.label) : host.valuePos(d.value);
+    }
+
+    yPos(host: XYAxis, d) {
+        return host.orientation() === "horizontal" ? host.valuePos(d.value) : host.dataPos(d.label);
+    }
+
+    layerUpdate(host: XYAxis, element, duration: number = 250) {
+        super.layerUpdate(host, element, duration);
+
+        element.selectAll("g").remove();
+
+        const data = this.flattenData();
+        const contourData = d3ContourDensity()
+            .x(d => this.xPos(host, d))
+            .y(d => this.yPos(host, d))
+            .size([this.width(), this.height()])
+            .bandwidth(this.contourBandwidth())
+            (data)
+            ;
+        const lines = element.insert("g", "g")
+            .attr("fill", this.contourFillColor())
+            .attr("stroke", this.contourStrokeColor())
+            .attr("stroke-linejoin", "round")
+            .selectAll("path")
+            .data(contourData)
+            ;
+        lines.enter().append("path")
+            .attr("d", geoPath())
+            ;
+        const points = element.append("g")
```

Comment:
This append (if its needed,) should be done in layerEnter (so it only happens once).